App Builder experience for Configurable Logging

buskamuza opened the pull request:
Added commands for managing log forwarding configuration.

- added `app:config:get:log-forwarding` command
- added `app:config:set:log-forwarding` command
- added `app:config:get:log-forwarding:errors` command
- updated log forwarding config on server during `app:deploy` command
- updated `app:logs` command to display info message in case custom log forwarding is configured
- save log forwarding settings to project config

## Description

See https://github.com/adobe/aio-cli-plugin-app/issues/501 and https://jira.corp.adobe.com/browse/DEVX-1244 (internal)

## Related Issue

- https://github.com/adobe/aio-cli-plugin-app/issues/501
- https://github.com/adobe/aio-lib-runtime/pull/83
- https://github.com/adobe/aio-cli-plugin-runtime/pull/254
- https://github.com/adobe/aio-lib-runtime/pull/93

## Motivation and Context

See https://github.com/adobe/aio-cli-plugin-app/issues/501

## How Has This Been Tested?

Automated and manual testing.

## Screenshots (if appropriate):

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

purplecabbage commented on src/lib/log-forwarding.js:
```diff
@@ -0,0 +1,253 @@
+  }
+
+  getLocalConfig () {
+    const config = this.aioConfig.project.workspace.log_forwarding
```
can this be null?

buskamuza replied:
If somebody writes it as `null` then yes. I'm checking for `undefined` now. But we can add more checks to take into account cases when people change the config manually and put something unexpected.

purplecabbage replied:
you are checking for undefined, what I mean though is that this code will throw if workspace, or project, or aioConfig is null/undefined.
if we were node@14+ then you could use optional chaining 
`const config = this.aioConfig?.project?.workspace?.log_forwarding`

buskamuza replied:
I haven't seen "parent" parts to be null and I checked with other code and it doesn't check these levels for null/undefined. But I can do it.

purplecabbage commented on src/lib/log-forwarding.js:
```diff
@@ -0,0 +1,253 @@
+      destination = destinations[0]
+      settings = configJson[destination]
+    } else {
+      throw new Error(`Configuration has ${destinations.length} destinations. Exactly one must be defined.`)
```
I am not sure why this check is required, how would we get a config object with multiple destinations?  Seems like that would be an error in code.

buskamuza replied:
If somebody changes config manually. We can't guarantee this doesn't happen, so there should be intentional behavior in this case. But in normal scenario this shouldn't happen.

purplecabbage commented on src/lib/log-forwarding.js:
```diff
@@ -0,0 +1,253 @@
+      destinationSettings.forEach(e => {
+        if (e.type === SECRET_FIELD_TYPE) {
+          const secretVarName = getSecretVarName(destination, e.name)
+          if (process.env[secretVarName] !== undefined) {
```
this is very mysterious ... who sets this secret? how? why process.env?

buskamuza replied:
I looked at an examples like this one https://github.com/adobe/aio-lib-core-config/blob/master/src/Config.js#L51 
Is there a better way to read values from `.env` file?

buskamuza replied:
Though I agree that this code depends on the `dotenv(true)` happening before it (which happens during config load). So if there is a better way (a clear API for reading secret data from `.env` or whatever place is expected for this data), let's use it.

purplecabbage commented on src/lib/log-forwarding.js:
```diff
@@ -0,0 +1,253 @@
+    }
+  }
+
+  getSupportedDestinations () {
```
nit: 1 line becomes 3 lines, just call this.logForwarding.getSupportedDestinations() directly

buskamuza replied:
Yes. I was thinking about those two ways. The only reason for the current implementation is fewer dependencies in the actual command. Right now, it depends only on this library. If we remove this method, it will have to depend also on the RT LogForwarding directly. I chose one of the options as it didn't look critical for me and this library looks just like a wrapper. But if it's a concern, I can remove these two methods.

purplecabbage replied:
my mistake, didn't see that this was called from outside

purplecabbage commented on src/lib/log-forwarding.js:
```diff
@@ -0,0 +1,253 @@
+
+  async updateLocalConfig (lfConfig) {
+    const destination = lfConfig.getDestination()
+    const destinationSettings = this.logForwarding.getDestinationSettings(destination)
```
nit: we defined `getSettingsConfig` 5 lines ago ... why aren't we using it?
just kidding: this line is correct, I would remove the `getSettingsConfig` function all together, just like `getSupportedDestinations` above.

purplecabbage commented on src/lib/log-forwarding.js:
```diff
@@ -0,0 +1,253 @@
+  }
+}
+
+class LogForwardingConfig {
```
is this worth an object/class?

buskamuza replied:
IMO, client code (commands) look better now. But this is probably objective.

purplecabbage commented on src/lib/log-forwarding.js:
```diff
@@ -0,0 +1,253 @@
+  }
+
+  isLocalConfigChanged () {
+    if (fs.pathExistsSync(process.cwd() + '/tmp/' + CHECKSUM_FILE)) {
```
- Use the dist/ folder for this, which is part of aioConfig
- this code is not windows friendly either
- process.cwd() is not needed
- use path.join instead of adding path strings yourself

buskamuza replied:
Updated

purplecabbage reviewed (changes requested):
mostly nitpicky stuff, but this seems to be a lot more code than required ... more code, more bugs

Deploy can be called with `--[no-]actions` in which case we should not be doing anything with log-forwarding.

buskamuza commented on package.json:
```diff
@@ -23,6 +23,7 @@
     "chalk": "^4.0.0",
     "chokidar": "^3.5.2",
     "cli-ux": "^5.4.5",
+    "crypto": "^1.0.1",
```
Removed

buskamuza commented:
> mostly nitpicky stuff, but this seems to be a lot more code than required ... more code, more bugs
> 
> Deploy can be called with `--[no-]actions` in which case we should not be doing anything with log-forwarding.

Updated

purplecabbage commented:
a couple issues I have found.
- when setting with `https://` the server does not return an `https://` url so `aio app config get log-forwarding` always shows a mismatch
- changing type via `aio app config set log-forwarding` overwrites values in .aio but leaves keys in .env

buskamuza commented:
> changing type via aio app config set log-forwarding overwrites values in .aio but leaves keys in .env

This was discussed with Manik. It leaves keys in `.env` if you change destination type. If you use the same destination type the key is overwritten. So far we agreed that this is expected behavior.

buskamuza commented:
> when setting with https:// the server does not return an https:// url so aio app config get log-forwarding always shows a mismatch

Fixed. Updated Runtime API to return sanitized settings and aio CLI then stores this instead of what the user provided. Secrets are not exchanged and stored as provided by the user. See also new PR https://github.com/adobe/aio-lib-runtime/pull/93 necessary to cover this.
